### PR TITLE
Restrict filesystem creation if path referred either '.' or '..'

### PR DESCRIPTION
**Pull request by TulsiJain**
Restrict filesystem creation if given path contains either '.' or '..'

### Motivation and Context

https://github.com/zfsonlinux/zfs/issues/8564

### Description

This changes restrict filesystem creation if given path contains either '.' or '..'
### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: include/zfs_namecheck.h [hunk 1]
```diff
@@ -63,4 +65,4 @@ int permset_namecheck(const char *, namecheck_err_t *, char *);
 }
 #endif
 
-#endif	/* _ZFS_NAMECHECK_H */
+#endif	/* _ZFS_NAMECHECK_H */
```
Comment (richardelling) [inline include/zfs_namecheck.h]: please add back the [return] here

File: include/zfs_namecheck.h [hunk 2]
```diff
@@ -43,6 +43,8 @@ typedef enum {
 	NAME_ERR_RESERVED,		/* entire name is reserved */
 	NAME_ERR_DISKLIKE,		/* reserved disk name (c[0-9].*) */
 	NAME_ERR_TOOLONG,		/* name is too long */
+	NAME_ERR_SELF_MOUNT,	/* pool mount is not allowed */
```
Comment (richardelling) [inline include/zfs_namecheck.h]: in the comment, pools cannot be mounted. I think you meant "filesystem"
Perhaps use
```
/* filesystem mount onto self is not allowed */
/* filesystem mount onto parent is not allowed */
```

File: lib/libzfs/libzfs_dataset.c [hunk 3]
```diff
@@ -196,6 +196,16 @@ zfs_validate_name(libzfs_handle_t *hdl, const char *path, int type,
 				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 				    "reserved disk name"));
 				break;
+			
+			case NAME_ERR_SELF_MOUNT:
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "self reference is found"));
```
Comment (richardelling) [inline lib/libzfs/libzfs_dataset.c]: I think this error message can be clearer. Perhaps "filesystem name references self"?

File: lib/libzfs/libzfs_dataset.c [hunk 4]
```diff
@@ -196,6 +196,16 @@ zfs_validate_name(libzfs_handle_t *hdl, const char *path, int type,
...
+				break;
+			
+			case NAME_ERR_PARENT_MOUNT:
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "parent path is found"));
```
Comment (richardelling) [inline lib/libzfs/libzfs_dataset.c]: To be more clear, perhaps "filesystem name references parent"?

File: module/zcommon/zfs_namecheck.c [hunk 5]
```diff
@@ -222,7 +222,9 @@ entity_namecheck(const char *path, namecheck_err_t *why, char *what)
 		}
 
 		/* Validate the contents of this component */
+		int component_length = 0;
```
Comment (richardelling) [inline module/zcommon/zfs_namecheck.c]: This seems to be too much work. Can this logic be replaced by a simpler strncmp()?

File: module/zfs/dsl_scan.c [hunk 6]
```diff
@@ -3025,8 +3025,10 @@ dsl_scan_async_block_should_pause(dsl_scan_t *scn)
 	if (zfs_recover)
 		return (B_FALSE);
 
-	if (scn->scn_visited_this_txg >= zfs_async_block_max_blocks)
+	if (zfs_async_block_max_blocks != 0 &&
```
Comment (richardelling) [inline module/zfs/dsl_scan.c]: This code change does not appear to be relevant to this PR?

Comment (TulsiJain) [reply module/zfs/dsl_scan.c]: Yes, this changes are of my previous PR which is accepted but not merged yet! I remove this for this now

File: include/zfs_namecheck.h (hunk 2 above)
Comment (TulsiJain) [reply include/zfs_namecheck.h]: I will replace with those /* self path reference in filesystem is not allowed */
/* parent path reference in filesystem is not allowed */

File: lib/libzfs/libzfs_dataset.c (hunk 3 above)
Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: How about NAME_ERR_SELF_REFERENCE and NAME_ERR_PARENT_REFERENCE

File: lib/libzfs/libzfs_dataset.c (hunk 4 above)
Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: How about NAME_ERR_PARENT_REFERENCE

File: lib/libzfs/libzfs_dataset.c (hunk 3 above)
Comment (richardelling) [reply lib/libzfs/libzfs_dataset.c]: I apologize for the confusion. I like the enum names. My comments referred to the comments or error message text.

File: module/zcommon/zfs_namecheck.c (hunk 5 above)
Comment (TulsiJain) [reply module/zcommon/zfs_namecheck.c]: I do not think so. We need to compare with every component. strncmp would not provide the desired.

File: lib/libzfs/libzfs_dataset.c (hunk 3 above)
Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: No problem, I feel they were not clear NAME_ERR_SELF_REFERENCE and NAME_ERR_PARENT_REFERENCE makes more sense.

File: include/zfs_namecheck.h [hunk 7]
```diff
@@ -43,6 +43,8 @@ typedef enum {
 	NAME_ERR_RESERVED,		/* entire name is reserved */
 	NAME_ERR_DISKLIKE,		/* reserved disk name (c[0-9].*) */
 	NAME_ERR_TOOLONG,		/* name is too long */
+	NAME_ERR_SELF_REFERENCE,	/* self path reference in filesystem is not allowed */
```
Comment (richardelling) [inline include/zfs_namecheck.h]: lines are limited to 80 characters with 8-character tabs. I think we can elide "is not allowed"

Comment (TulsiJain) [reply include/zfs_namecheck.h]: will fix it thanks

File: lib/libzfs/libzfs_dataset.c [hunk 8]
```diff
@@ -196,6 +196,16 @@ zfs_validate_name(libzfs_handle_t *hdl, const char *path, int type,
 				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 				    "reserved disk name"));
 				break;
+			
+			case NAME_ERR_SELF_REFERENCE:
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "self reference is found"));
```
Comment (ahrens) [inline lib/libzfs/libzfs_dataset.c]: I think it might be easier for admins to understand a more specific error message, like `"dot ('.') component in name"`.  You could check the error message in context to see how it flows, e.g.:
```
$ zfs create test/.
bla bla bla
self reference is found
```

File: lib/libzfs/libzfs_dataset.c [hunk 9]
```diff
@@ -196,6 +196,16 @@ zfs_validate_name(libzfs_handle_t *hdl, const char *path, int type,
...
+				break;
+			
+			case NAME_ERR_PARENT_REFERENCE:
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "parent reference is found"));
```
Comment (ahrens) [inline lib/libzfs/libzfs_dataset.c]: same here:  `"dot-dot ('..') component in name"`

Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: parent reference, ('..') is found in name

File: lib/libzfs/libzfs_dataset.c (hunk 8 above)
Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: how about? self reference, ('.') is found in name

File: module/zcommon/zfs_namecheck.c [hunk 10]
```diff
@@ -222,7 +222,9 @@ entity_namecheck(const char *path, namecheck_err_t *why, char *what)
...
 		for (const char *loc = start; loc != end; loc++) {
+			component_length++;
```
[the remark this answers is not in the document]
Comment (TulsiJain) [reply module/zcommon/zfs_namecheck.c]: Yes we can do that. And, I should also put one more if condition around that end should refer to either  '/' or '\0'.

File: module/zcommon/zfs_namecheck.c [hunk 11]
```diff
@@ -232,6 +234,24 @@ entity_namecheck(const char *path, namecheck_err_t *why, char *what)
...
+			if (*start == '.') {
+				if (why)
+					*why = NAME_ERR_SELF_REFERENCE;
+				return (-1);
+			}
+		}
+
+		/* Validate the contents of this component is not '..'*/
+		if (component_length == 2){
+			if (*start == '.' && start[1] == '.') {
```
[the remark this answers is not in the document]
Comment (TulsiJain) [reply module/zcommon/zfs_namecheck.c]: Ok corrected!

**Comment by TulsiJain** [comment]
> one more style issue.
> Also, I'd like to see the test added so we can cover the various cases:
> 
> * testpool/.
> * testpool/..
> * testpool/something/./else
> * testpool/this/../that
>   This will be easy to add to the list of negative tests in `tests/zfs-tests/tests/functional/cli_root/zfs_create/zfs_create_009_neg.ksh`

I have added tests cases.

File: lib/libzfs/libzfs_dataset.c [hunk 12]
```diff
@@ -196,6 +196,16 @@ zfs_validate_name(libzfs_handle_t *hdl, const char *path, int type,
 				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 				    "reserved disk name"));
 				break;
+			
+			case NAME_ERR_SELF_REFERENCE:
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "self reference, ('.') is found in name"));
```
Comment (richardelling) [inline lib/libzfs/libzfs_dataset.c]: I don't think we need two enclosures for the reference, can we do without the parenthesis?

Comment (TulsiJain) [reply lib/libzfs/libzfs_dataset.c]: Done!

File: include/zfs_namecheck.h [hunk 13]
```diff
@@ -43,6 +43,8 @@ typedef enum {
 	NAME_ERR_RESERVED,		/* entire name is reserved */
 	NAME_ERR_DISKLIKE,		/* reserved disk name (c[0-9].*) */
 	NAME_ERR_TOOLONG,		/* name is too long */
+	NAME_ERR_SELF_REFERENCE,	/* self path reference in filesystem*/
+	NAME_ERR_PARENT_REFERENCE,	/* parent path reference in filesystem*/
```
Comment (behlendorf) [inline include/zfs_namecheck.h]: nit: Like the `NAME_ERR_DISKLIKE` case (and printed error messages) I think it would be very helpful to include the '.', and ',,' strings in these comments.  What do you think about the following.

```c
/* reserved self path name ('.') */
/* reserved parent path name ('..') */
```

File: module/zcommon/zfs_namecheck.c [hunk 14]
```diff
@@ -182,7 +182,7 @@ int
 entity_namecheck(const char *path, namecheck_err_t *why, char *what)
 {
 	const char *end;
-
+
```
Comment (behlendorf) [inline module/zcommon/zfs_namecheck.c]: nit: stray whitespace change

File: include/zfs_namecheck.h (hunk 13 above)
Comment (TulsiJain) [reply include/zfs_namecheck.h]: OK

File: module/zcommon/zfs_namecheck.c (hunk 14 above)
Comment (TulsiJain) [reply module/zcommon/zfs_namecheck.c]: ok

**Comment by behlendorf** [comment]
@TulsiJain I believe all that's still needed here to squash the commits and force update the PR so we can get a clean run of the CI.  You'll need to run something like this:

```
git checkout restrict_mount
git rebase -i master

# In your editor of choice replace 'pick' with 'squash' for all but the first
# commit then exit.  For each commit which is squashed keep only the
# final commit message.  After the rebase and squash you will have a
# single commit to force push.

git push --force TulsiJain restrict_mount
```

**Comment by TulsiJain** [comment]
> @TulsiJain I believe all that's still needed here to squash the commits and force update the PR so we can get a clean run of the CI. You'll need to run something like this:
> 
> ```
> git checkout restrict_mount
> git rebase -i master
> 
> # In your editor of choice replace 'pick' with 'squash' for all but the first
> # commit then exit.  For each commit which is squashed keep only the
> # final commit message.  After the rebase and squash you will have a
> # single commit to force push.
> 
> git push --force TulsiJain restrict_mount
> ```

Done

**Comment by behlendorf** [comment]
Thanks.  It looks like there are just a few more minor touch ups needed to address the [code style warnings](http://build.zfsonlinux.org/builders/Ubuntu%2017.10%20x86_64%20%28STYLE%29/builds/14157/steps/shell_2/logs/stdio).  You can run `make checkstyle` locally to reproduce the warnings and verify they're fixed before refreshing the PR.

**Comment by TulsiJain** [comment]
> Thanks. It looks like there are just a few more minor touch ups needed to address the [code style warnings](http://build.zfsonlinux.org/builders/Ubuntu%2017.10%20x86_64%20%28STYLE%29/builds/14157/steps/shell_2/logs/stdio). You can run `make checkstyle` locally to reproduce the warnings and verify they're fixed before refreshing the PR.

Done, Thanks for helping me learn!

File: include/zfs_namecheck.h [hunk 15]
```diff
@@ -43,6 +43,8 @@ typedef enum {
 	NAME_ERR_RESERVED,		/* entire name is reserved */
 	NAME_ERR_DISKLIKE,		/* reserved disk name (c[0-9].*) */
 	NAME_ERR_TOOLONG,		/* name is too long */
+	NAME_ERR_SELF_REFERENCE,	/* reserved self path name ('.') */
+	NAME_ERR_PARNT_REFERENCE,	/* reserved parent path name ('..') */
```
Comment (behlendorf) [inline include/zfs_namecheck.h]: Sorry, I missed this initially.  `NAME_ERR_PARNT_REFERENCE` should be `NAME_ERR_PARENT_REFERENCE`.

Comment (TulsiJain) [reply include/zfs_namecheck.h]: Unfortunately, It crosses that 80 words limit in  `module/zcommon/zfs_namecheck.c` line 250. That why I had to change it to `NAME_ERR_PARNT_REFERENCE`. Should I go for another look around?

Comment (richardelling) [reply include/zfs_namecheck.h]: REF is often used as shorthand for REFERENCE. So try NAME_ERR_PARENT_REF

Comment (ahrens) [reply include/zfs_namecheck.h]: Please use the correct spelling.  You can break line 250 after the `=`.

Comment (TulsiJain) [reply include/zfs_namecheck.h]: Corrected!
